operator/src/main/java/org/keycloak/operator/upgrade/impl/BaseUpgradeLogic.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright 2025 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.operator.upgrade.impl;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+
+import io.fabric8.kubernetes.api.model.Container;
+import io.fabric8.kubernetes.api.model.apps.StatefulSet;
+import io.fabric8.zjsonpatch.JsonDiff;
+import io.javaoperatorsdk.operator.api.reconciler.Context;
+import io.javaoperatorsdk.operator.api.reconciler.UpdateControl;
+import io.quarkus.logging.Log;
+import org.keycloak.operator.ContextUtils;
+import org.keycloak.operator.controllers.KeycloakDeploymentDependentResource;
+import org.keycloak.operator.crds.v2alpha1.CRDUtils;
+import org.keycloak.operator.crds.v2alpha1.deployment.Keycloak;
+import org.keycloak.operator.upgrade.UpgradeLogic;
+import org.keycloak.operator.upgrade.UpgradeType;
+
+/**
+ * Common {@link UpgradeLogic} implementation that checks if the upgrade logic needs to be run.
+ * <p>
+ * The upgrade logic can be skipped if it is the first deployment or if the change is not relevance (like, updating the
+ * number of replicas or annotations).
+ */
+abstract class BaseUpgradeLogic implements UpgradeLogic {
+
+    // JSON path key in the diff.
+    private static final String PATH = "path";
+    // Container fields that may trigger an upgrade.
+    private static final Collection<String> FIELDS = List.of("/env", "/envFrom", "/image");
+
+    protected final Context<Keycloak> context;
+    protected final Keycloak keycloak;
+    protected final KeycloakDeploymentDependentResource statefulSetResource;
+
+    BaseUpgradeLogic(Context<Keycloak> context, Keycloak keycloak, KeycloakDeploymentDependentResource statefulSetResource) {
+        this.context = context;
+        this.keycloak = keycloak;
+        this.statefulSetResource = statefulSetResource;
+    }
+
+    @Override
+    public final Optional<UpdateControl<Keycloak>> decideUpgrade() {
+        var existing = context.getSecondaryResource(StatefulSet.class);
+        if (existing.isEmpty()) {
+            // new deployment, no upgrade needed
+            Log.debug("New deployment - skipping upgrade logic");
+            return Optional.empty();
+        }
+        var desiredStatefulSet = statefulSetResource.desired(keycloak, context);
```

Comment:
Just a non-blocking nitpick. I'm not sure if we manually need to call `desired` here. Possibly it's possible we could fetch it somewhere from the context as `desired` is called by the JOSDK itself. CC @shawkins

Reply:
Ok, thanks for the confirmation. I'm resolving this.